File: Solutions/Python/n_queens_ii.py
"""
    [HARD]
    52. N-Queens II

    Concepts:
    - backtracking

    Stats:
        Runtime | 7 ms      [Beats 89.33%]
        Memory  | 17.66 MB  [Beats 88.49%]
"""

class Solution:
    def totalNQueens(self, n: int) -> int:
        count = 0
        used_cols = set()
        diags = set()
        antidiags = set()

        def backtrack(i: int) -> None:
            nonlocal count

            if i == n:
                count += 1
                return

            for j in range(n):
                if j in used_cols or (i + j) in diags or (i - j) in antidiags:
                    continue

                used_cols.add(j)
                diags.add((i + j))
                antidiags.add((i - j))

                backtrack(i + 1)

                used_cols.remove(j)
                diags.remove((i + j))
                antidiags.remove((i - j))

        backtrack(0)
        return count

File: Solutions/Cpp/n_queens_ii.cpp
/*
    [HARD]
    52. N-Queens II

    Concepts:
    - backtracking

    Stats:
        Runtime | 8 ms      [Beats 13.05%]
        Memory  | 10.53 MB  [Beats 9.69%]
*/

class Solution {
public:
    int totalNQueens(int n) {
        int count = 0;
        std::unordered_set<int> used_cols = {};
        std::unordered_set<int> diags = {};
        std::unordered_set<int> antidiags = {};

        std::function<void(int)> backtrack;
        backtrack = [&](int i){
            if (i == n) {
                count++;
                return;
            }

            for (int j = 0; j < n; j++) {
                if (
                    used_cols.contains(j) ||
                    diags.contains(i + j) ||
                    antidiags.contains(i - j)
                ) {
                    continue;
                }

                used_cols.insert(j);
                diags.insert(i + j);
                antidiags.insert(i - j);

                backtrack(i + 1);

                used_cols.erase(j);
                diags.erase(i + j);
                antidiags.erase(i - j);
            }

            return;
        };

        backtrack(0);
        return count;
    }
};